opendatahub-io/data-science-pipelines
# feat: add artifact storage api for backend

Part of https://issues.redhat.com/browse/RHOAIENG-8225

**Description of your changes:**
This change introduces a new api for artifact data retrieval for mlpipeline api server, to be consumed by UI and end users. Two new endpoints are introduced, allowing for 1) collection retrieval and 2) individual artifact metadata retrieval.

When an individaul artifact resource is retrieved, users/ui can specify a share_url option that provides a pre-signed url that with a very short expiry date that can be used as an artifact download link directly from the originating object store provider. This is currently only implemented for s3-compliant object store backends.

Related upstream issue: https://github.com/kubeflow/pipelines/issues/10872

## Testing instructions

(try the following with Minio, and AWS S3) 

1. Deploy DSPA  (set `enableExternalRoute: true`  if using default minio via dspa in objectstorage field example [here](https://github.com/opendatahub-io/data-science-pipelines-operator/pull/670/files#diff-83d5649c2166cd22649899e7c45c37da112253e15affd9cc10f679e962e690e4R12))
2. Run an Iris pipeline
3. Fetch a single artifact with GET `{{dsp-route}}/apis/v2beta1/artifacts/1?share_url=true`   the "share_url=true" parameter will provide a signed url link that will last for 15 seconds (configurable via `SIGNED_URL_EXPIRY_TIME_SECONDS` env var). Evaluate the response object and confirm it matches with the artifact in mlmd. 
4. Click the share_url (within 15 seconds) or paste in browser, confirm it allows you to download the artifact.
5. Fetch a collection of artifacts with a GET on `{{dsp-route}}/apis/v2beta1/artifacts?namespace=YOUR-DSPA-NS` evaluate the response object. Note that `share_url` is not a field here. 
6. Try other parameters for fetch collection, like `max_result_size` which limits the number of artifacts returned, it will also return a `next_page_token` which can be fed into another subsequent GET request on the same endpoint but with a `next_page_token` parameter. 
7. For fetching collection confirm `is_asc` (true or false) will sort by ascending or non-ascending respectively.
8. `order_by_field` will order by the following fields: FIELD_UNSPECIFIED, CREATE_TIME, LAST_UPDATE_TIME, ID

**Checklist:**
